File: 06-arrays-sorting/01_two_sum.py
# ============================================================
# PATTERN: ARRAYS SORTING
# ============================================================
# 1. **Two pointers** -- sorted arrays, pair/triplet sum, container problems
# 2. **Prefix sum** -- range sum queries, subarray sum problems
# 3. **Hash map** -- O(1) lookup for complement, frequency counting
# 4. **Merge intervals** -- sort by start, merge overlapping
# 5. **Dutch National Flag** -- 3-way partition with lo/mid/hi pointers
#
# ---
# ============================================================

import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


# Two Sum (LC #1) -- Easy
# Find two numbers that add up to target. Return indices.
#
#   Key: Hash map: for each num, check if `target - num` exists

def two_sum(nums, target):
    """Return indices of two numbers that add up to target"""
    pass


# Order doesn't matter, so sort result
def solve(nums, target):
    r = two_sum(nums, target)
    return sorted(r) if r else r


run_tests(solve, [
    (([2,7,11,15], 9),   [0, 1]),
    (([3,2,4], 6),       [1, 2]),
    (([3,3], 6),         [0, 1]),
])

File: 06-arrays-sorting/01_two_sum.c
#include "../test_runner.h"

/* Write result into out[0], out[1] */
void two_sum(int* nums, int n, int target, int* out) {
    out[0] = -1; out[1] = -1;
}

int main() {
    int r[2];
    two_sum(ARR(2,7,11,15), 4, 9, r); TEST_ARR("basic",  r, ARR(0,1), 2);
    two_sum(ARR(3,2,4), 3, 6, r);     TEST_ARR("mid",    r, ARR(1,2), 2);
    two_sum(ARR(3,3), 2, 6, r);       TEST_ARR("dupes",  r, ARR(0,1), 2);
    TEST_SUMMARY();
}
